tools/bridge/bridge/event_fetcher.py

```diff
@@ -70,36 +70,30 @@ def fetch_events_in_range(
         else:
             self.logger.debug(f"Found {len(events)} events.")
 
-        for event in events:
-            self.event_queue.put(event)
+        return events
 
-    def fetch_events_not_seen(self) -> None:
-        self.logger.debug("Fetch new events.")
-        self.logger.debug(
-            f"Last fetched block number: {self.last_fetched_block_number}."
-        )
+    def fetch_some_events(self) -> List:
+        """fetch some events starting from the last_fetched_block_number
 
-        fetch_until_block_number = max(
-            self.web3.eth.blockNumber - self.max_reorg_depth, 0
-        )
+        This method tries to fetch from consecutive ranges of blocks
+        until it has found some events in a range of blocks or it has
+        reached the head of the chain.
 
-        self.logger.debug(
-            f"Fetch until {fetch_until_block_number} respecting "
-            f"the maximum reorg depth of {self.max_reorg_depth}."
-        )
-
-        for from_block_number in range(
-            self.last_fetched_block_number + 1,
-            fetch_until_block_number + 1,
-            self.event_fetch_limit,
-        ):
+        This method returns an empty list if the caller should wait
+        for new blocks to come in.
+        """
+        while True:
```

Comment:
I think this loop is a little hard to understand. If I understand correctly, it is only necessary because we use the empty list to denote that the caller should wait, so we can't use it to denote that there are no events. Which is a bit confusing to me, because the latter seems like the much more appropriate use for an empty list.

Have you thought about getting rid of the loop and either returning `None` or raising some kind of exception if the chain hasn't progressed since the last call? I think this would make the code a little easier to understand.

Reply:
> I think this loop is a little hard to understand. If I understand correctly, it is only necessary because we use the empty list to denote that the caller should wait, so we can't use it to denote that there are no events

it actually denotes that the caller should wait *and* that there are no new events. Raising an exception is bad because it's not an exceptional situation. Returning None only makes it harder to handle the return value.

Reply:
it's a bit like the read function returning an empty string at EOF. To me this is a common idiom.